tools/unirestore/scylla_restore.py

```diff
@@ -0,0 +1,996 @@
+            procs.append([p, cmd])
+        while success and len(procs) > 0:
+            for proc in procs:
+                if proc[0].poll() == None: # Still running
+                    pass
+                elif proc[0].poll() == 0:  # Successfully exited
+                    out, err = proc[0].communicate()
+                    logger.debug('completed: stdout: %s stderr: %s', out, err)
+                    procs.remove(proc)
+                    if len(err.splitlines()) > 0:
+                        logger.error('failed on rsync %s with %s', out, err)
+                        success = False
+                else:                      # Failed
+                    success = False
+                    try:
+                        logger.error('failed %s :: %s', proc[1], str(proc[0].communicate()))
+                    except Exception as e:
+                        logger.error('failed on rsync', exc_info=True)
+                        # exit is in the calling func because delete_failed_bundle() is required
+        if not success:
+        # kill off all the other rsync processes if one failed
+            for p in procs:
+                logger.warning('Killing leftover rsync process for %s', str(p[1]))
+                p[0].kill()
+        return success
+    if 'pkey' in ssh_auth.keys():
+        sshopts = ('-e "ssh -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null '
+                   '-o LogLevel=ERROR -i %s"' % ssh_auth['pkey'])
+    elif 'password' in ssh_auth.keys():
+        sshopts = ('-e "sshpass -p %s ssh -o StrictHostKeyChecking=no '
+                   '-o UserKnownHostsFile=/dev/null -o LogLevel=ERROR"' % ssh_auth['password'])
+    else:
+        logger.error('Invalid ssh credentials in configuration, exiting')
+        sys.exit(1)
+    success = True
+    procs = []
+    if not bwlimit or bwlimit == 'None':
+        rsync = ('rsync -az --rsync-path=\"sudo /usr/bin/rsync\"')
+    else:
+        rsync = ('rsync -az --bwlimit=%s --rsync-path=\"sudo /usr/bin/rsync\"' % str(bwlimit))
+    if int(copy_procs) >= len(node_list) or int(copy_procs) == 0:
+        success = run_rsync(file_path=file_path, node_list=node_list, dest_path=dest_path,
+                            bwlimit=bwlimit, copy_procs=copy_procs, ssh_auth=ssh_auth)
+    else:
+        node_list = [node_list[i:i + copy_procs] for i in range(0, len(node_list), copy_procs)]
+        logger.debug('splitting rsync processes into %s size chunks: %s', str(copy_procs), str(node_list))
+        for sublist in node_list:
+            _success = run_rsync(file_path, sublist, dest_path, bwlimit, copy_procs, ssh_auth)
+            if not _success:
+                success = False
+                return success
+                # return early. exit is in the calling func because delete_failed_bundle() is required
+    return success
+
+# def build_restore_plan(file_list, ring, dest_dirs, src_nodes, dest_nodes):
```

Comment:
this comment is still needed ?

Reply:
This is part of the "hachana lemazgan" for the token-aware restore. Not very aesthetic, I agree

Reply:
I thought no need for that in Canada 😂
I guess you'll write/test that part when needed.

Reply:
I even tested it, it works. But sstablemetadata is VERY slow and I had to kick it out until there's a faster approach for reading tokens out of sstable files. I have a small python script that does it quickly (provided by @raphaelsc ), but keeping and maintaining it within the scope of this tool would have been wrong.